### ISSUE TITLE
about grpo loss computation

### ISSUE THREAD

**Issue by gameofdimension** [volcengine/verl#593]
https://github.com/volcengine/verl/blob/cae8d2f5f65e9b2a94aa0b31eab4d40120eade6e/verl/trainer/ppo/core_algos.py#L301

for grpo, this line should be modified to 

```python
pg_loss = verl_F.masked_mean(torch.max(pg_losses, pg_losses2), eos_mask, axis=1).mean()
```

1. the grpo paper says we should first average between tokens within one sequence, then average between sequences
2. `pg_loss = verl_F.masked_mean(batch, eos_mask, axis=1).mean()` **not equal to** `pg_loss = verl_F.masked_mean(batch, eos_mask)`, a simple example:

```
import torch
import verl.utils.torch_functional as verl_F

batch = torch.tensor(
    [
        [2,  2, 2, 2, 2, 6, 5, 3, 2, 4],
        [10, 3, 4, 5, 2, 3, 4, 1, 1, 3]
    ]
).type(torch.float32)

eos_mask = torch.tensor(
    [
        [1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]
)

v1 = verl_F.masked_mean(batch, eos_mask)
v2 = verl_F.masked_mean(batch, eos_mask, axis=1).mean()

print(v1, v2)
```

output

> tensor(3.3333) tensor(6.)

**Comment by LLouice** [comment]
@gameofdimension Good catch!

How to choose normalization is indeed a point worth considering and it's easy to overlook.

I prefer to decide the actual difference between sequence-level normalization and global normalization through comparative experiments.

@gameofdimension @tongyx361 What are your thoughts on this issue?

**Comment by tongyx361** [comment]
Indeed nice catch! We are planning to examine the effect of different weighting through experiments and theory analysis, also providing comprehensive configurations at the framework level. Please stay tuned!
